cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,956 @@
+            raise ValueError(
+                'Datasets must be populated before adding stages. '
+                'Provide `input_datasets`, or omit `stages_in_order` and call '
+                'pipeline.set_stages(stages_in_order) later.'
+            )
+
+        self.name = name
+        self.version = version or time.strftime('%Y%m%d-%H%M%S')
+        self.namespace = namespace
+
+        if storage_policy == StoragePolicy.CPG:
+            storage_provider = CpgStorageProvider(cloud)
+        else:
+            raise PipelineError(f'Unsupported storage policy {storage_policy}')
+
+        self.cohort = Cohort(
+            self.name,
+            analysis_dataset_name=analysis_dataset,
+            namespace=namespace,
+            storage_provider=storage_provider
+        )
+        self.refs = RefData(storage_provider.get_ref_bucket())
+
+        self.check_intermediates = check_intermediates
+        self.check_expected_outputs = check_expected_outputs
+        self.skip_samples_with_missing_input = skip_samples_with_missing_input
+        self.first_stage = first_stage
+        self.last_stage = last_stage
+
+        self.tmp_bucket = to_path(
+            self.cohort.analysis_dataset.get_tmp_bucket(
+                version=f'{self.name}/{self.version}'
+            )
+        )
+        self.keep_scratch = keep_scratch
+        self.dry_run: bool = dry_run
+
+        if local_dir:
+            self.local_dir = to_path(local_dir)
+            self.local_tmp_dir = to_path(tempfile.mkdtemp(dir=local_dir))
+        else:
+            self.local_dir = self.local_tmp_dir = to_path(tempfile.mkdtemp())
+
+        self.prev_batch_jobs = dict()
+        if previous_batch_tsv_path is not None:
+            assert previous_batch_id is not None
+            self.prev_batch_jobs = PrevJob.parse(
+                previous_batch_tsv_path,
+                previous_batch_id,
+                tmp_bucket=self.tmp_bucket,
+                keep_scratch=keep_scratch,
+            )
+
+        self.config = config or {}
+
+        if version:
+            description += f' {version}'
+        if input_datasets:
+            description += ': ' + ', '.join(input_datasets)
+        self.b: Batch = setup_batch(
+            description=description, 
+            tmp_bucket=self.tmp_bucket,
+            keep_scratch=self.keep_scratch,
+            billing_project=self.cohort.analysis_dataset.stack,
+        )
+
+        self.status_reporter = None
+        input_provider = None
+        if (
+            input_provider_type == InputProviderType.SMDB 
+            or status_reporter_type == StatusReporterType.SMDB
+        ):
+            smdb = SMDB(self.cohort.analysis_dataset.name)
+            if status_reporter_type == StatusReporterType.SMDB:
+                self.status_reporter = SmdbStatusReporter(smdb)
+            if input_provider_type == InputProviderType.SMDB:
+                input_provider = SmdbInputsProvider(smdb)
```

Comment:
I don't have a strong opinion about this, but you might prefer a design where the `Pipeline` class itself is completely agnostic to the underlying concrete implementations of the providers and only depends on the abstract interfaces. This way you avoid accidentally depending on concrete implementation details of any of the providers.

You'd then need to add a pipeline factory helper function (which can be completely isolated from the pipeline implementation itself), which instantiates those concrete providers and passes them to the pipeline.

Reply:
Good idea. Refactored it!

Reply:
Added a function `create_pipeline` (that can be run now with `cli_opts` instead of `Pipeline`), and simplified the `Pipeline` constructor to take already initialised `Cohort`, `Batch`, `StatusReporter`, and other objects. Let me know if it looks good.